### ISSUE TITLE
Custom toolchain for cross compilation

### ISSUE THREAD

**Issue by abhinav-radius** [nokeedev/gradle-native#367]
I am trying to define a custom toolchain (using a cross compiler) for some C code that I want to compile as linux aarch64 from an x64 MacOSX build machine . I have some cross compile toolchain installed on my build machine from https://github.com/thinkski/osx-arm-linux-toolchains. I am struggling to understand how to define this custom toolchain. Is this possible? If yes, can someone point me to the build.gradle changes I would need to make. I am not well-versed in gradle so forgive me if this is something that is trivially known.

**Comment by lacasseio** [comment]
Correct, it all boils down to the target for the toolchain. In the example, we configure the target with this line: `toolchain.target("macosarm64", new ConfigurePlatformAction(serviceRegistry.get(ExecActionFactory.class), toolchain, sdkPath));` Notice the `macosarm64` string. Nokee uses the concatenation of the os name and arch name all in lower case. In the example, the OS name is `macOS` and the architecture name is `arm64` so the result string is `macosarm64`. If you were to use a target machine like: `machines.linux.architecture('arm64')` then the target string would be `linuxarm64`.

Everything else is standard Gradle core features. You can read more about the `toolChains` container [here](https://docs.gradle.org/current/userguide/native_software.html#native_binaries:tool_chain).

**Comment by lacasseio** [comment]
Ah yes, you are right on that part. Some of that debt should be cleaned up in the next 4 weeks. If you have multiple targets for a component you could invoke the assemble lifecycle task directly and it should build. Unfortunately, it won't connect to the JAR task. Single target components are strangely optimized which makes the compile/link task "hidden" from the CLI point of view. This will be fixed soon.

It may be possible to change the logic in that exact point to check if it's buildable instead of using the OS family matching. I will try to find some time today to try it out.

**Comment by abhinav-radius** [comment]
Will give this a try sometime today and let you know. In the meantime, another issue that will have to be solved are the JNI headers. Currently, it is picked based on host OS. I think , these will have to be picked using target machine and some toolchain for the target machine. 

```
private void registerJvmHeaderSourceSet(JavaNativeInterfaceLibrary extension) {
		// TODO: This is an external dependency meaning we should go through the component dependencies.
		//  We can either add an file dependency or use the, yet-to-be-implemented, shim to consume system libraries
		//  We aren't using a language source set as the files will be included inside the IDE projects which is not what we want.
		val jvmIncludes = getJvmIncludes();
		extension.getBinaries().configureEach(BaseNativeBinary.class, binary -> {
			binary.getCompileTasks().configureEach(NativeSourceCompileTask.class, task -> {
				((AbstractNativeCompileTask) task).getIncludes().from(jvmIncludes);
			});
		});
	}

	private Provider<List<File>> getJvmIncludes() {
		return getProviders().provider(() -> {
			List<File> result = new ArrayList<>();
			result.add(new File(Jvm.current().getJavaHome().getAbsolutePath() + "/include"));

			if (OperatingSystem.current().isMacOsX()) {
				result.add(new File(Jvm.current().getJavaHome().getAbsolutePath() + "/include/darwin"));
			} else if (OperatingSystem.current().isLinux()) {
				result.add(new File(Jvm.current().getJavaHome().getAbsolutePath() + "/include/linux"));
			} else if (OperatingSystem.current().isWindows()) {
				result.add(new File(Jvm.current().getJavaHome().getAbsolutePath() + "/include/win32"));
			}
			return result;
		});
	}
```

**Comment by abhinav-radius** [comment]
> I made a quick fix to how the buildability of a variant is determined, see #385. I released version `0.4.556-202110111448.5620125b` containing the fix. Could you give it a try?
> 
> The fix won't go in master soon-ish simply because it impacts the other native components/variants which need to follow the same convention and they don't use the same code path for the buildability. However, you can use that version for the time been. The path was done on the latest master.

This change worked for me. Thanks for being so responsive.

**Comment by lacasseio** [comment]
No worries, I'm sure the fix should be enough for your need.
